File: src/main/scala/design_patterns_with_scala/chapter01/Bookshelf.scala
package design_patterns_with_scala.chapter01

case class Book(name: String)

class Bookshelf(books: Book*) extends Iterable[Book]:
  def iterator: Iterator[Book] = books.iterator

@main
def bookshelfExample() =
  // Iterable
  val bookshelf = new Bookshelf(
    Book("Around the World in 80 Days"),
    Book("Bible"),
    Book("Cinderella"),
    Book("Daddy-Long-Legs"),
  )
  for book <- bookshelf do println(book.name)

File: src/main/clj/design_patterns_with_clojure/chapter01/bookshelf.clj
(ns design-patterns-with-clojure.chapter01.bookshelf)

(defrecord Book [name])

(deftype Bookshelf
  [books]
  clojure.lang.Seqable
  (seq [_]
    (seq books)))

(defn -main []
  ;; Seqable
  (let [bookshelf (->Bookshelf
                   [(->Book "Around the World in 80 Days")
                    (->Book "Bible")
                    (->Book "Cinderella")
                    (->Book "Daddy-Long-Legs")])]
    (doseq [book bookshelf]
      (println (:name book)))))
